ecleankernel/file.py

```diff
@@ -101,18 +142,38 @@ def read_internal_version(self) -> str:
             raise UnrecognizedKernelError(
                 f'Kernel file {self.path} terminates before bzImage '
                 f'header')
-        if buf[2:6] != b'HdrS':
-            raise UnrecognizedKernelError(
-                f'Unmatched magic for kernel file {self.path} '
-                f'({repr(buf[2:6])} != b"HdrS")')
-        offset = struct.unpack_from('H', buf, 0x0e)[0]
-        f.seek(offset - 0x10, 1)
-        buf = f.read(0x100)  # XXX
-        if not buf:
+        if buf[2:6] == b'HdrS':
+            offset = struct.unpack_from('H', buf, 0x0e)[0]
+            f.seek(offset - 0x10, 1)
+            buf = f.read(0x100)  # XXX
+            if not buf:
+                raise UnrecognizedKernelError(
+                    f'Kernel file {self.path} terminates before expected '
+                    f'version string position ({offset + 0x200})')
+            ret = buf.decode(errors='ignore').split(' ', 1)
+        else:
+            # If it's not a bzImage it must be a raw binary,
+            # check if it's compressed first
+            b = self.__decompress_raw()
+            # unlike with bzImage, the raw kernel binary has no header
+            # that includes the version, so we parse the version message
+            # that appears on boot
+            s = b.decode('ascii', 'ignore')
+            ver_start = 'Linux version '
+            pos = s.find(ver_start)
+            if pos == -1:
+                raise UnrecognizedKernelError(
+                    f'Kernel file {self.path} does not appear '
+                    f'to have a version string, '
+                    f'or the compression format was not recognized')
+            pos = pos + len(ver_start)
+            sbuf = s[pos:pos + 0x100]
```

Comment:
Is there a reason for this `0x100`? In the above code, the idea was not to have to read the whole file into memory but you do that already anyway.

Reply:
Now that we only decode as needed, there is indeed a reason for it :)